### PR TITLE
fix default_plugin warnings with recent dart sdk

### PR DESCRIPTION
**Pull request by Sameri11**
This PR addresses #1555 and removes warning. Implementation is based on [this comment](https://github.com/flutter/flutter/issues/152037#issuecomment-2239909463) from https://github.com/flutter/flutter/issues/152037

Fixes https://github.com/miguelpruivo/flutter_file_picker/issues/1484
Fixes https://github.com/miguelpruivo/flutter_file_picker/issues/1483
Fixes https://github.com/miguelpruivo/flutter_file_picker/issues/1343

**Details:**  

Old implementation with `default_plugin` never did anything over than providing correct list of supported platforms for package page on pub.dev. Since some version of dart sdk, this behaviour has become warning.
New implementation provides (arguably) correct usage of `flutter:plugins:` section to avoid warning. Since all logic for choosing current platform and appropriate plugin's implementation is encapsulated inside current code, this PR just provides empty `registerWith` static methods needed for compilation. Also, it is mandatory for plugin implementations to be public, so we export them.

**Testing:**

Current tests are not affected by this PR. Example app is also functioning (tested on `android` and `ios`). Since tests on macOS (my dev machine) and example on both emulators works – I believe that functionality is not broken.

#1555 provides reproduction with https://github.com/zulip/zulip-flutter repo. This fix can be tested in that repo by adding `dependency_overrides` to `pubspec.yaml`. This override will point to my fork with fix:

```yaml
dependency_overrides:
  file_picker:
    git:
      url: git@github.com:Sameri11/flutter_file_picker.git
      ref: fix-issue-1555
```

### CODE REVIEW

**Comment by Sameri11** [comment]
Hey @miguelpruivo, @philenius can you please review this fix and/or approve workflow?

File: lib/src/file_picker_io.dart [hunk 1]
```diff
@@ -16,6 +16,8 @@ const EventChannel _eventChannel =
 
 /// An implementation of [FilePicker] that uses method channels.
 class FilePickerIO extends FilePicker {
+  static void registerWith() {}
```
Comment (navaronbracke) [inline lib/src/file_picker_io.dart:19]: Don't you have to set `FilePicker.platform` to an instance of `FilePickerIO()` in `registerWith()` ?
(and the other implementations in the other registerWith() functions)

You might also need to double check the one for web.

Then you could get rid of the `filePickerWithFFI()` stub and fix a bunch of issues relating to using the `Platform` class on the web.

The following issues are caused by that bug: 
https://github.com/miguelpruivo/flutter_file_picker/issues/1484
https://github.com/miguelpruivo/flutter_file_picker/issues/1483
https://github.com/miguelpruivo/flutter_file_picker/issues/1343

If you could get that fixed in this PR also, that would be great, since it has been a bit of a pain to deal with (and I didn't have time to look into fixing it myself yet)

**Review by navaronbracke — CHANGES_REQUESTED** [review]
Mainly a comment to see if we can fix a bunch of long standing issues with mis-use of `Platform._operatingSystem` in the plugin.

**Comment by OlaiyaO** [comment]
I see you're all actively working on it.
I'm still experiencing the issue here. 

"Package file_picker:windows references file_picker:windows as the default plugin, but it does not provide an inline implementation.
Ask the maintainers of file_picker to either avoid referencing a default implementation via `platforms: windows: default_package: file_picker` or add an inline implementation to file_picker via `platforms: windows:` `pluginClass` or `dartPluginClass`."

**Comment by navaronbracke** [comment]
> I see you're all actively working on it. I'm still experiencing the issue here.
> 
> "Package file_picker:windows references file_picker:windows as the default plugin, but it does not provide an inline implementation. Ask the maintainers of file_picker to either avoid referencing a default implementation via `platforms: windows: default_package: file_picker` or add an inline implementation to file_picker via `platforms: windows:` `pluginClass` or `dartPluginClass`."

That is indeed what this PR would aim to solve yes. Once we get the requires changes in this PR, I'll review it again and then we can release a fix version.

File: lib/src/file_picker_io.dart (hunk 1 above)
Comment (Sameri11) [reply lib/src/file_picker_io.dart:19]: Thanks for your review and additional clarification! My goal here was to provide minimal effort fix for the warning, hence this dummy implementation. 
I would try to make changes to address your comment soon, thanks.

Comment (Sameri11) [reply lib/src/file_picker_io.dart:19]: > Don't you have to set FilePicker.platform to an instance of FilePickerIO() in registerWith() ?

About this – as far as I understand, it is the best practice :) So yes, I guess we can move logic there and it is preferred option. I would try that.
But current variant works because all logic for platform implementation choosing is already in _setPlatform(), so we needed only dummy empty method to satisfy compiler.

Comment (Sameri11) [reply lib/src/file_picker_io.dart:19]: Well, I dived in this issue and I don't think that I'm happy with what I found :)

When platform implementation is added as `dartPluginClass`, this implementation must be exported from main library file, as per [dart_plugin_registrant template](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/flutter_plugins.dart#L621-L673)

In our case, it means that `lib/file_picker.dart` will look something like this for registrant to be able to resolve implementation classes:

```dart
library file_picker;

export './src/file_picker.dart';
export './src/platform_file.dart';
export './src/file_picker_result.dart';

// Added lines start from here

export './src/file_picker_macos.dart';
export './src/linux/file_picker_linux.dart';
export './src/file_picker_io.dart';
export './src/windows/file_picker_windows.dart';
```

Of course it will break in web build on last line when exporting windows implementation for very same reason (ffi types) which `filePickerWithFFI()` needed for. Basically, this abstraction just goes to another level, where, instead of stub implementation, stub export is needed.

I will investigate further, but at this point I'm not sure that some appropriate solution, without creation of federation plugins, is possible. 

Would love to hear another opinions.

Comment (navaronbracke) [reply lib/src/file_picker_io.dart:19]: According to https://docs.flutter.dev/packages-and-plugins/developing-packages#dart-only-platform-implementations

you don't specifically have to export these? Exporting an implementation would only be needed if there is extra functionality exposed by the specific implementation (i.e. an extra method that only exists for that implementation)

Currently the plugin does not expose such "platform specific extra functionality". If that is something we want to provide, we should switch to providing sub-packages for each platform.

And if we absolutely need a conditional export, you could do the following:

1) In https://github.com/miguelpruivo/flutter_file_picker/blob/master/lib/src/windows/file_picker_windows.dart change the existing stub to return an instance of `FilePickerWindows`

```
FilePickerWindows filePickerWithFFI() => FilePickerWindows();
```

2) Use the existing stub from https://github.com/miguelpruivo/flutter_file_picker/blob/master/lib/src/windows/stub.dart
as the stub implementation for the conditional export.

Comment (Sameri11) [reply lib/src/file_picker_io.dart:19]: > you don't specifically have to export these? 

I've also read this part of docs and thought like that too. But there is implementation detail which is not throughly described in documentation: framework generates `dart_plugin_registrant.dart` file where it calls `registerWith()` methods of concrete implementation. After `pubspec.yaml` modifications file looks like this:

```dart
//
// Generated file. Do not edit.
// This file is generated from template in file `flutter_tools/lib/src/flutter_plugins.dart`.
//

// @dart = 2.19

import 'dart:io'; // flutter_ignore: dart_io_import.
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';

@pragma('vm:entry-point')
class _PluginRegistrant {

  @pragma('vm:entry-point')
  static void register() {
    if (Platform.isAndroid) {
      try {
        FilePickerIO.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isIOS) {
      try {
        FilePickerIO.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isLinux) {
      try {
        FilePickerLinux.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isMacOS) {
      try {
        FilePickerMacOS.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isWindows) {
      try {
        FilePickerWindows.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    }
  }
}
```

As you can see, it only imports `package:file_picker/file_picker.dart`. And if I'm not wrong, there is no possibility to change this behaviour.


As for conditional import – I considered these options too. But I believe that some kind of better implementation is possible :) Thing is, while breaking web build, those exported members are never really used there. That's why I think that stub implementation is a little bit overkill and, maybe, just export a completely empty stub file is more appropriate.

But, again, I will investigate further and come back with some solution shortly.

And, of course, thanks for your guidance and feedback!

Comment (stuartmorgan) [reply lib/src/file_picker_io.dart]: In the short term, you will indeed need to either do a conditional import layer (`path_provider_windows` and `path_provider_linux` have two different variants of that approach) to make the implementation class web-safe, or use the stub `defaultClass` approach.

Comment (Sameri11) [reply lib/src/file_picker_io.dart]: @stuartmorgan thanks!

**Comment by Sameri11** [comment]
@navaronbracke can you please take another look at this? I think I implemented what you talked about [here](https://github.com/miguelpruivo/flutter_file_picker/pull/1567#discussion_r1702126247).

Btw, I had hard times reproducing #1343 both based on this PR and master – no success. Either I do something wrong, or, maybe, these issues were fixed earlier somehow.

File: lib/file_picker.dart [hunk 2]
```diff
@@ -3,3 +3,9 @@ library file_picker;
 export './src/file_picker.dart';
 export './src/platform_file.dart';
 export './src/file_picker_result.dart';
+export './src/file_picker_macos.dart';
+export './src/linux/file_picker_linux.dart';
+export './src/file_picker_io.dart';
+// Conditional export needed for web to successfully compile.
+export './src/windows/file_picker_windows_stub.dart'
+    if (dart.library.ffi) './src/windows/file_picker_windows_real.dart';
```
Comment (navaronbracke) [inline lib/file_picker.dart:11]: Can we keep the `file_picker_windows.dart` filename for the real implementation? (rather than `file_picker_windows_real.dart`)
The fact that the fake one is called `file_picker_windows_stub.dart` clarifies its use in my opinion.

File: lib/file_picker.dart [hunk 3]
```diff
@@ -3,3 +3,9 @@ library file_picker;
 export './src/file_picker.dart';
 export './src/platform_file.dart';
 export './src/file_picker_result.dart';
+export './src/file_picker_macos.dart';
+export './src/linux/file_picker_linux.dart';
+export './src/file_picker_io.dart';
+// Conditional export needed for web to successfully compile.
```
Comment (navaronbracke) [inline lib/file_picker.dart:9]:
```suggestion
// Conditional export needed for web to successfully compile,
// as `dart:ffi` is not available on the web.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,8 @@
+## 8.0.7
+### General
```
Comment (navaronbracke) [inline CHANGELOG.md:2]: The last bullet point isn't needed, since that is an implementation detail.
Also, perhaps we should reword it to something like this?
I also added a specific note for people who were using a workaround for the Platform class issue like:

```
if (kIsWeb) {
  FilePickerWeb()
} else {
  FilePicker()
}
```

```suggestion
### General
- Fixes an issue relating to incorrect registration of platform-specific implementations. [#1555](https://github.com/miguelpruivo/flutter_file_picker/issues/1555).
- Fixes an issue that required users to use `kIsWeb` and `FilePickerWeb` when on the web.
```

**Comment by navaronbracke** [comment]
I also edited the original PR comment to list fixed issues, so these get closed when the PR is merged.

As for not being able to reproduce https://github.com/miguelpruivo/flutter_file_picker/issues/1343 I'm not sure either.

File: lib/file_picker.dart (hunk 2 above)
Comment (Sameri11) [reply lib/file_picker.dart]: Have no strong opinion about this. So yes, done :)

File: CHANGELOG.md (hunk 4 above)
Comment (Sameri11) [reply CHANGELOG.md:2]: Fully agree here, thanks!

**Review by navaronbracke — APPROVED** [review]
LGTM!

I also edited out the redundant bullet points (one was an implementation detail, which users don't care about and the other was now a duplicate point since we already mention fixing #1555 in the first point)

**Comment by navaronbracke** [comment]
It looks one of the test is failing, because https://github.com/miguelpruivo/flutter_file_picker/pull/1561/files did not update a test expectation. I'll address this with a commit and then this can land!

**Comment by navaronbracke** [comment]
Checks are green, so ship it!
